ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,749 @@
+				Err(EngineError::RequiresSigner)?
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => Ok((sig, digest)),
+					Err(e) => Err(EngineError::Custom(e.into()))?,
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		debug_assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		self.block_state_by_hash.write().get_mut(hash).cloned()
+	}
+
+	/// Get `CliqueBlockState` for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		let mut block_state_by_hash = self.block_state_by_hash.write();
+		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+			return Ok(state.clone());
+		}
+		// If we are looking for an checkpoint block state, we can directly reconstruct it.
+		if header.number() % self.epoch_length == 0 {
+			let state = self.new_checkpoint_state(header)?;
+			block_state_by_hash.insert(header.hash(), state.clone());
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
+		match self.client.read().as_ref().and_then(|w| w.upgrade()) {
+			None => {
+				return Err(EngineError::RequiresClient)?;
+			}
+			Some(c) => {
+				let last_checkpoint_number = header.number() - header.number() % self.epoch_length as u64;
+				debug_assert_ne!(last_checkpoint_number, header.number());
+
+				let mut chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity(
+					(header.number() - last_checkpoint_number + 1) as usize);
+
+				// Put ourselves in.
+				chain.push_front(header.clone());
+
+				// populate chain to last checkpoint
+				loop {
+					let (last_parent_hash, last_num) = {
+						let l = chain.front().expect("chain has at least one element; qed");
+						(*l.parent_hash(), l.number())
+					};
+
+					if last_num == last_checkpoint_number + 1 {
+						break;
+					}
+					match c.block_header(BlockId::Hash(last_parent_hash)) {
+						None => {
+							return Err(BlockError::UnknownParent(last_parent_hash))?;
+						}
+						Some(next) => {
+							chain.push_front(next.decode()?);
+						}
+					}
+				}
+
+				// Catching up state, note that we don't really store block state for intermediary blocks,
+				// for speed.
+				let backfill_start = time::Instant::now();
+				trace!(target: "engine",
+						"Back-filling block state. last_checkpoint_number: {}, target: {}({}).",
+						last_checkpoint_number, header.number(), header.hash());
+
+				// Get the state for last checkpoint.
+				let last_checkpoint_hash = *chain.front()
+											 .expect("chain has at least one element; qed")
```

Comment:
Indentation looks really weird here.

Reply:
It's a Github issue with tab indentation (it appears at 8 spaces instead of 4).  Looks a lot better in an editor IMO

Reply:
I have fixed nvm, it was a space that messed it up :P